proposals/p2511.md

```diff
@@ -0,0 +1,190 @@
+    can be inlined, at the cost of additional work for the compiler.
+-   In C++, overloading an operator `$` does not automatically provide a
+    matching `$=`, resulting in additional work or incomplete operator sets.
+
+## Proposal
+
+C-family assignment operators are provided as statements:
+
+-   `variable = value;` is a simple assignment statement.
+-   `variable $= value;` is a compound assignment for each binary operator `$`,
+    other than comparisons. `<=` and `>=` mean "less than or equal to" and
+    "greater than or equal to", not "compare and assign".
+-   `++variable;` and `--variable;` are supported as increment and decrement
+    syntax. Because these are statements, there is no distinction between pre-
+    and post-increment, and post-increment is not provided.
+
+These operations are translated into calls on interfaces.
+
+## Details
+
+See the changes to the design.
+
+This proposal does not define the semantics of assignment that are provided for
+classes by default. Leads issue
+[#686](https://github.com/carbon-language/carbon-lang/issues/686) gives some
+rules, but those rules are not part of this proposal.
+
+## Rationale
+
+-   [Language tools and ecosystem](/docs/project/goals.md#language-tools-and-ecosystem)
+    -   The values of variables can only change if either the address is taken,
+        including implicitly by `addr` self parameters, or at assignment
+        statements, making it easier to reason about where the value of a
+        variable can change in the control flow of a function.
+-   [Software and language evolution](/docs/project/goals.md#software-and-language-evolution)
+    -   This approach is conservative and can evolve to support assignment as a
+        subexpression.
+-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
+    -   Assignments in subexpressions tend to be hard for humans to read and
+        understand. Disallowing them at the language level avoids the potential
+        for confusion, making code easier to read at the cost of making certain
+        constructs such as `n = arr[i++];` a little harder to write.
+    -   Easier to write a complete operator set because both `$` and `$=` can be
+        provided by implementing a single interface.
+-   [Practical safety and testing mechanisms](/docs/project/goals.md#practical-safety-and-testing-mechanisms)
+    -   Replaces correctness warning on `if (a = b)` with a language rule.
+-   [Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
+    -   Providing largely the same set of symbols makes migration easier.
+    -   Some cost may be imposed by forcing a translation from assignment in
+        subexpressions to assignment as separate statements.
+
+## Alternatives considered
+
+### Allow assignment as a subexpression
```

Comment:
I will note particular concerns with this for increment and decrement. e.g., [absl's flat_hash_map erase](https://github.com/abseil/abseil-cpp/blob/master/absl/container/flat_hash_map.h#L229-L233) relies on postincrement as a subexpression. This could make particular challenges for migrating code, as either adding additional locals (since `map.erase(it); ++it;` is potentially different) or creating new APIs becomes a prerequisite for migration, likely reducing automated migration.

But it's not clear to me that you're looking for this kind of concern at this time, so I'll mostly suggest that if you're still considering this, "alternatives considered" indicates that it has been ruled out; "open questions" might be more appropriate? I think the difference here is whether the plan is "look more" versus "wait and see if there are too many complaints"; if it's the latter, I would agree with "alternatives considered", but would also want more consideration for how we migrate existing C++ code on the assumption that this isn't reconsidered.

Reply:
Thought: for migration, maybe we just transform C++ `it++` into something like `[&]() -> auto { var x: auto = it; ++it; return x; }()`

Reply:
Thanks, I've added some brief discussion here giving the rationale for rejecting this alternative, which was previously missing, and also describing possibilities for getting the same effect (eg, for migration) and some potential limitations when doing so.

Reply:
What do you think about having a distinct alternative for increment and decrement? It can I think be much more brief, but I feel like there are some unique aspects to those, both in terms of the impact and potentially more limited pitfalls. And so it might be clearer to treat those separately.

Minor point either way though -- there is a fundamental tradeoff here in that this will create some unknown amount of migration friction, and it'll be experience with tooling and actual migrations that teach us whether we end up want to revisit parts of this or not.

Reply:
Just to be explicit, I'm fine with the resolution here, although apprehensive about migration costs.